feat: add Twig functions and tokens with inheritance alternatives

JoshuaBehrens opened the pull request:
### 1. Why is this change necessary?

There are plugins out there, that make use of the Twig function `block` to include another's template block. This is nice to reuse code but does not check the inheritance and therefore fails to include other changes. The plugin author will be requested to use this code afterwards.

### 2. What does this change do, exactly?

I copied/reused functions or token parsers from Twig and made sure, they are checking the inheritance so you use all Twig ways to build your templates with all benefits of the plugin system.

### 3. Describe each step to reproduce the issue or behaviour.

* Add `sw_block` function as alternative to `block` but with inheritance
* Add `sw_source` function as alternative to `source` but with inheritance
* Add `sw_include` function as alternative to `sw_include` token
* Add `sw_use` token as alternative to `use` but with inheritance
* Add `sw_from` token as alternative to `from` but with inheritance
* Add `sw_import` token as alternative to `import` but with inheritance
* Add `sw_embed` token as alternative to `embed` but with inheritance
* Use `sw_block` in templates, which results in same cache but as a better example

### 4. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes ( https://github.com/shopware/docs/issues/1500 )
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

JoshuaBehrens committed:
Add sw_block function as alternative to block but with inheritance

Add sw_source function as alternative to source but with inheritance
Add sw_include function as alternative to sw_include token
Add sw_use token as alternative to use but with inheritance
Add sw_from token as alternative to from but with inheritance
Add sw_import token as alternative to import but with inheritance
Add sw_embed token as alternative to embed but with inheritance
Use sw_block in templates, which results in same cache but as a better example

BrocksiNet commented:
Hi, @JoshuaBehrens :wave: First of all, great PR, and thx for that contribution!

Currently, the Storefront Team is aiming for less inheritance or, in other words, well-defined places for inheritance because the community also complained about too many breaks in the storefront area. This is also why we have the discussion about the [Storefront Future](https://github.com/shopware/shopware/discussions/4856).

So, from an extension point of view, this change makes a lot of sense. I am just wondering about the maintenance and if this will complicate things. And what should I use when I, as an extension developer, want to create a block without any inheritance? We need to test this a lot, explain it in Docs, and make sure that more TWIG functions do not make the system slower.

We will come back to you. Will take some time. Anyway, good work. :muscle:

JoshuaBehrens commented:
@BrocksiNet thank you for the feedback.

At first: the reason why I made this was an update in a plugin from a well-known plugin manufacturer, that referenced an OG Shopware file using the Twig block function. So some people are not aware, that not everything is inheritance aware. I shared my playground from the hacktoberfest event https://gist.github.com/JoshuaBehrens/f528e643adcd8c3b0a306b3bb7a44088 to verify things work.

I noticed the discussion and I feel like sw_embed would fit right in there. So I do not see it contrary to it but rather supporting that point.

I added some tests so I hope there are maybe just some other code path tests missing, that you can add.

I added some docs with reference to the Symfony docs. So I do not think we are missing any documentation. With the point some parts of your public communication claim: when you know Symfony, you have a headstart in working with Shopware. But when Twig is only covered partially it is rather a hassle than helpful. So I think this is a missing step we had to take.

BrocksiNet commented:
@JoshuaBehrens, I have prepared everything and have some internal approvals.

We want to merge this with 6.7 because we can't foresee what will change for a shop due to these new inheritances. 

Is that okay with you? Or do you not see any problems or breaks that could occur?

JoshuaBehrens commented:
@BrocksiNet I intended my work to behave without breaking changes I see your concerns though. If we can have a separate feature flag, I am fine with that :) because I need it :D 

And please make sure my name is on the commit. sounds weird but in the recent past some cherry-picks or however you (as a company) take the commits over were just with the person on your side and one feels to be less part of the project especially with big changes like this :/

BrocksiNet commented:
Added the [new PR](https://github.com/shopware/shopware/pull/5869) with all the changes needed. Currently as Draft.

JoshuaBehrens commented:
@BrocksiNet I already checked "Allow edits and access to secrets by maintainers" how shall we proceed? I offer to rebase :) anything else?

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/NodeExtension.php:
```diff
@@ -16,6 +16,8 @@ class NodeExtension extends AbstractExtension
     /**
      * @internal
+     *
+     * @deprecated tag:v6.7.0  - replace TemplateFinder with TemplateFinderInterface
```
```suggestion
     * @deprecated tag:v6.8.0  - replace TemplateFinder with TemplateFinderInterface
```

it is too late for 6.7 :grin:

BrocksiNet replied:
Oh no :scream: Not a big problem I guess ... this can wait

BrocksiNet commented:
In my opinion, this is okay now. 

@JoshuaBehrens, could you test again because of the twig update to see if everything works as expected?

I do not trust the `codecov` number, to be honest ... -15.18% sounds like a lot ... any suggestions? :face_with_diagonal_mouth:

keulinho commented on src/Core/Framework/Adapter/Twig/TokenParser/EmbedTokenParser.php:
```diff
@@ -0,0 +1,77 @@
+ * @see \Twig\TokenParser\EmbedTokenParser
+ */
+#[Package('framework')]
+final class EmbedTokenParser extends \Twig\TokenParser\IncludeTokenParser
```
Should we mark all of those classes as internal? Just to be free in the future to react to changes, this should be the case for everything in adapter/twig

BrocksiNet replied:
Done :white_check_mark:

BrocksiNet committed:
NEXT-38820 - Introduce sw_use specific adapter exception

BrocksiNet committed:
NEXT-38820 - Move unit test to own file

BrocksiNet committed:
NEXT-38820 - TemplateFinder instead of TemplateFinderInterface

BrocksiNet committed:
NEXT-38820 - update to new twig version 3.19.0

BrocksiNet committed:
chore: Change package from core to framework

BrocksiNet committed:
NEXT-38820 - Introduce sw_block specific adapter exception

BrocksiNet committed:
chore: TokenParser internal

JoshuaBehrens committed:
Change EmbedTokenParser to copy more code to not depend on internal code

JoshuaBehrens commented:
@BrocksiNet I rebased and sqashed some commits/parts of commits to cleanup the history a bit. After re-reviewing my own code, I saw, that the EmbedTokenParser depends on an internal class from Twig. I changed it by copying over that specific piece of code.

There are some issues around bc-checks, but they were introduced by you adding internal. Is this something irrelevant as we are only having major changes on trunk?

You already updated the code to match new twig code. We have tests. We have a(n improved 💙) changelog. Anything that is missing out?

BrocksiNet committed:
Merge branch 'trunk' into feature/sw-block-twig-function

BrocksiNet committed:
feat: ignore bc error

mitelg commented on .bc-exclude.php:
```diff
@@ -50,6 +50,13 @@
         // major
         'Value of constant Shopware\\\\Core\\\\Kernel::SHOPWARE_FALLBACK_VERSION changed from \'6.6.9999999-dev\' to \'6.7.9999999-dev\'',
 
+        // Class is marked as @internal (can be removed after merge to 6.7)
```
do we still need that? bc checker is currently disabled. if we want to deprecate something, this needs to be done for 6.8

mitelg commented on changelog/_unreleased/2024-10-04-add-twig-functions-and-tokens-with-their-alternatives-with-inheritance.md:
```diff
@@ -0,0 +1,25 @@
+  * new Twig token `sw_import` with `\Shopware\Core\Framework\Adapter\Twig\TokenParser\ImportTokenParser`
+___
+# Storefront
+* Added Twig function `sw_block`, that must be used, when the template reference is used and inheritance is expected
```
do we want to deprecate the usage of the default Twig tokens and functions? :thinking: 

do we need to ensure, that only the new ones are used?

BrocksiNet replied:
What would be the benefit of forcing this?

mitelg replied:
proper inheritance in all places? :man_shrugging: I just wanted to know, if this is not necessary, we can leave it like this

BrocksiNet replied:
The main question is: Does the developer know that when he uses just `block` he only has a single inheritance? Maybe this is what he wants, and that would be okay with me. Or is he just using block because he copied that somewhere from Twig docs and does not know the difference between `sw_block` (multi inheritance)?

We need to explain the differences, but we should not favor one over the other. And, of course, we cannot prevent some from being used with the wrong intention due to a lack of knowledge.

mitelg replied:
sure :+1: I just thought it could be confusing, because sometimes you can and sometimes you have to use `sw_*` stuff :slightly_smiling_face: (e.g. sw_extends)

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -80,6 +82,25 @@ public static function invalidTemplateSyntax(string $message): self
+    {
+        return new self(
+            Response::HTTP_BAD_REQUEST,
+            self::INVALID_ARGUMENT,
```
if it is the `INVALID_ARGUMENT` code, you should use the already existing exception method. if not, define a new code

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/TwigFeaturesWithInheritanceExtension.php:
```diff
@@ -0,0 +1,118 @@
+use Twig\Util\CallableArgumentsExtractor;
+
+#[Package('framework')]
+class TwigFeaturesWithInheritanceExtension extends AbstractExtension
```
please mark all the new classes as internal

BrocksiNet replied:
![image](https://github.com/user-attachments/assets/16358721-f3d0-4a5b-b3ba-d6db1f6627e0)

Hmmm my local file looks different 🤔

keulinho replied:
is internal now

BrocksiNet committed:
feat: revert bc errors and internals

BrocksiNet committed:
feat: use invalidArgument

keulinho commented on src/Core/Framework/Adapter/Twig/TokenParser/MacroOverrideTokenParserMacro.php:
```diff
@@ -5,15 +5,13 @@
 
-#[Package('framework')]
 /**
- * @internal
```
why do we remove the internal here?

BrocksiNet replied:
I guess BC Checker was unhappy about that 🤔 But I can add and exclude it again.

mitelg replied:
we should make sure, that every class in the twig adapter namespace is internal. in general all classes in the adapter namespace should be internal

BrocksiNet committed:
feat: add internal to all TokenParser

BrocksiNet committed:
Merge branch 'trunk' into feature/sw-block-twig-function

BrocksiNet committed:
feat: updated UPGRADE-6.7.md

BrocksiNet committed:
Merge branch 'trunk' into feature/sw-block-twig-function

mitelg commented:
thanks for your contribution @JoshuaBehrens 👍🎉💙

JoshuaBehrens commented:
Thank you for working it out together :)

nguyenquocdaile commented:
Hi @JoshuaBehrens , For the changes you have suggested in the commercial plugin when replacing `block` to `sw_block`
```
{% block swag_advanced_search_suggest_result_link %}
	{{ block('layout_search_suggest_result_link', '@Storefront/storefront/layout/header/search-suggest.html.twig') }}
{% endblock %}
```
to
```
{% block swag_advanced_search_suggest_result_link %}
	{{ sw_block('layout_search_suggest_result_link', '@Storefront/storefront/layout/header/search-suggest.html.twig') }}
{% endblock %}
```
We found an issue with the rendering HTML element
![image](https://github.com/user-attachments/assets/17096395-cbed-4cd4-85aa-3af5906db27a)

So the solution here is to use a raw filter if we use sw_block
```
{% block swag_advanced_search_suggest_result_link %}
	{{ sw_block('layout_search_suggest_result_link', '@Storefront/storefront/layout/header/search-suggest.html.twig') |raw }}
{% endblock %}
```

Just raise it here when someone has met the issue.
Thanks

JoshuaBehrens commented:
@nguyenquocdaile thank you for point out :)
